domain/controllerconfig/state/state.go

```diff
@@ -56,44 +68,96 @@ func (st *State) ControllerConfig(ctx context.Context) (map[string]string, error
+	ctx context.Context,
+	updateAttrs map[string]string, removeAttrs []string,
+	validateModification func(map[string]string) error,
+) error {
 	db, err := st.DB()
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	selectQuery := "SELECT key, value FROM controller_config"
-	deleteQuery := "DELETE FROM controller_config WHERE key = ?"
-	updateQuery := `
-INSERT INTO controller_config (key, value)
-VALUES (?, ?)
-  ON CONFLICT(key) DO UPDATE SET value=?`
+	selectStmt, err := st.Prepare("SELECT &dbKeyValue.* FROM v_controller_config", dbKeyValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		// Check keys and values are valid between current and new config.
-		rows, err := tx.QueryContext(ctx, selectQuery)
+	deleteStmt, err := st.Prepare("DELETE FROM controller_config WHERE key IN ($S[:])", sqlair.S{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	removeKeys := sqlair.S(transform.Slice(removeAttrs, func(s string) any { return any(s) }))
+
+	var (
+		controllerStmt *sqlair.Statement
+		uuid           dbController
+	)
+	if v, ok := updateAttrs[controller.ControllerUUIDKey]; ok && v != "" {
+		controllerStmt, err = st.Prepare(`INSERT INTO controller (uuid) VALUES ($dbController.uuid)`, dbController{})
 		if err != nil {
 			return errors.Trace(err)
 		}
-		current, err := controllerConfigFromRows(rows)
-		if err != nil {
-			return errors.Trace(err)
+		uuid = dbController{UUID: v}
+	}
+
+	updateStmt, err := st.Prepare(`
+INSERT INTO controller_config (key, value)
+VALUES ($dbKeyValue.*)
+  ON CONFLICT(key) DO UPDATE SET value=excluded.value`, dbKeyValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	updateKeyValues := make([]dbKeyValue, 0)
+	for k, v := range updateAttrs {
+		if k == controller.ControllerUUIDKey {
+			continue
+		}
+		updateKeyValues = append(updateKeyValues, dbKeyValue{
+			Key:   k,
+			Value: v,
+		})
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Check keys and values are valid between current and new config.
+		var keyValues []dbKeyValue
+		if err := tx.Query(ctx, selectStmt).GetAll(&keyValues); err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return errors.Trace(err)
+			}
+		}
+
+		current := make(map[string]string)
+		for _, kv := range keyValues {
+			current[kv.Key] = kv.Value
 		}
 		if err := validateModification(current); err != nil {
 			return errors.Trace(err)
 		}
 
-		// Remove the attributes
-		for _, r := range removeAttrs {
-			if _, err := tx.ExecContext(ctx, deleteQuery, r); err != nil {
-				return errors.Trace(err)
+		// Insert the controller UUID, we ignore if it already exists, as
+		// the errors will tell use why it might not of worked.
+		if controllerStmt != nil {
+			if err := tx.Query(ctx, controllerStmt, uuid).Run(); err != nil {
+				if !internaldatabase.IsErrConstraintPrimaryKey(err) && !internaldatabase.IsErrConstraintUnique(err) {
```

Comment:
Let's not do this - just use an ON CONFLICT

Reply:
Sure.